File: backend/api/domain_kb.py
from fastapi import APIRouter, UploadFile, File, Form, HTTPException
from typing import List, Dict, Any
import pandas as pd
import io
from core.domain_kb.construct import stratified_sample, generate_rules_with_llm, save_rules, load_rules
from core.llm import call_llm  # Assume a unified LLM calling interface is available
import os

router = APIRouter()

@router.post("/domain_kb/sample")
def sample_data(
    file: UploadFile = File(...),
    frac: float = Form(0.1),
    min_rows: int = Form(100),
    stratify_cols: str = Form(None)
):
    """
    Upload a clean data file and return stratified sampling results (frontend preview).
    """
    try:
        df = pd.read_csv(file.file)
        cols = stratify_cols.split(',') if stratify_cols else None
        sample = stratified_sample(df, frac=frac, min_rows=min_rows, stratify_cols=cols)
        return {"columns": sample.columns.tolist(), "rows": sample.head(100).to_dict(orient="records")}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/domain_kb/generate_rules")
def generate_rules(
    file: UploadFile = File(...)
):
    """
    Input CSV file and directly generate rules.
    """
    try:
        df = pd.read_csv(file.file)
        # Extract table name from filename (remove .csv extension)
        table_name = file.filename.replace('.csv', '').replace('.CSV', '')
        rules = generate_rules_with_llm(df, call_llm, table=table_name)
        return {"rules": rules}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/domain_kb/save_rules")
def save_rules_api(
    rules: str = Form(...),
    dataset: str = Form(...)
):
    """
    Save reviewed rules in JSONL format.
    """
    try:
        import json
        # Parse rules from JSON string
        rules_list = json.loads(rules)
        if not isinstance(rules_list, list):
            raise ValueError("Rules must be an array")
        
        path = f"testdata/knowledge_base/domain_kb_{dataset}.jsonl"
        save_rules(rules_list, path)
        return {"status": "ok", "path": path, "count": len(rules_list)}
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="Invalid JSON format for rules")
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/domain_kb/get_rules")
def get_rules(dataset: str):
    """
    Get saved rules.
    """
    try:
        path = f"testdata/knowledge_base/domain_kb_{dataset}.jsonl"
        if not os.path.exists(path):
            raise HTTPException(status_code=404, detail="Rules not found")
        rules = load_rules(path)
        return {"rules": rules}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/api/domain_kb.js
// API calls for Domain Knowledge Base generation

export const sampleData = async (file, frac = 0.1, minRows = 100, stratifyCols = null) => {
  const uri = "http://localhost:8000/domain_kb/sample";
  const formData = new FormData();
  formData.append("file", file);
  formData.append("frac", frac);
  formData.append("min_rows", minRows);
  if (stratifyCols) {
    formData.append("stratify_cols", stratifyCols);
  }

  try {
    const data = await fetch(uri, {
      method: "POST",
      body: formData,
    }).then((res) => res.json());
    return data;
  } catch (error) {
    console.error("Error sampling data:", error);
    throw error;
  }
};

export const generateRules = async (fileOrFormData) => {
  const uri = "http://localhost:8000/domain_kb/generate_rules";
  try {
    const data = await fetch(uri, {
      method: "POST",
      body: fileOrFormData, // FormData with CSV file
    }).then((res) => res.json());
    return data;
  } catch (error) {
    console.error("Error generating rules:", error);
    throw error;
  }
};

export const saveRules = async (rules, dataset) => {
  const uri = "http://localhost:8000/domain_kb/save_rules";
  const formData = new FormData();
  formData.append("rules", JSON.stringify(rules));
  formData.append("dataset", dataset);

  try {
    const data = await fetch(uri, {
      method: "POST",
      body: formData,
    }).then((res) => res.json());
    return data;
  } catch (error) {
    console.error("Error saving rules:", error);
    throw error;
  }
};
